### PR TITLE
Add VI training options: num_particles, gradient clipping

### PR DESCRIPTION
**Pull request by shaharbar1**
### Changes:
* Add `num_particles`, `gradient_clip_norm` to VI `update_kwargs`
  - `num_particles` is passed to `TraceMeanField_ELBO`/`Trace_ELBO` for multi-particle gradient estimates
  - `gradient_clip_norm` chains `optax.clip_by_global_norm` before the base optimizer
* Switch all optimizers from numpyro-native to optax via `optax_to_numpyro`, removing the dual-path logic
* Remove `clipped_adam`, `momentum`, and `reduce_on_plateau` scheduler
* Remove `batch_size` parameter from `create_update_model`; read from `_update_kwargs` instead
* Add `optax` to `pyproject.toml` dependencies

## Summary by CodeRabbit

* **New Features**
  * Optax optimizers with selectable LR schedulers, optional global‑norm gradient clipping, and VI support for multiple particles; learning-rate and scheduler configurable via update options.

* **Breaking Changes**
  * Batch size is provided via update/training options; public model-construction API changed. Early stopping now uses absolute loss-change magnitude for convergence.

* **Tests**
  * Added/expanded tests for LR schedulers, VI options (particles, gradient clipping), categorical-feature updates, and Optax-kind consistency.

* **Chores**
  * Version bumped to 7.1.1; Optax added as a runtime dependency.

### CODE REVIEW

File: pybandits/model.py [hunk 1]
```diff
@@ -1454,7 +1521,42 @@ def __setstate__(self, state: dict) -> None:
...
+
+        Returns
+        -------
+        contextlib.AbstractContextManager
+            Either numpyro.handlers.scale(scale=beta) or nullcontext().
+        """
+        kl_tau = self._update_kwargs.get("kl_tau")
+        if kl_tau is not None:
+            n_neurons = max(sum(lp.weight.shape[1] for lp in self.model_params.bnn_layer_params[:-1]), 1)
+            return numpyro.handlers.scale(scale=kl_tau * n_samples / n_neurons)
```
Comment (ronshiff1) [inline pybandits/model.py]: since the assumption is that  n_samples / n_neurons << 1, maybe we can bound the entire scale by 1? 
meaning - this only decreases the KL weight?
can you please refer me to the results where they show that it works well for non overparameterized scenarios?

File: pybandits/model.py [hunk 2]
```diff
@@ -1486,31 +1583,36 @@ def create_update_model(self, batch_size: Optional[PositiveInt] = None) -> Calla
...
+            n_samples = x.shape[0]
 
             # Sample all weights (global parameters, outside plate)
             weights_biases = []
-            for layer_ind, layer_params in enumerate(self.model_params.bnn_layer_params):
-                weight_name, bias_name = self.get_layer_params_name(layer_ind)
-                w = numpyro.sample(weight_name, layer_params.weight.to_numpyro_distribution())
-                b = numpyro.sample(bias_name, layer_params.bias.to_numpyro_distribution())
-                weights_biases.append((w, b, layer_params.weight.shape))
+            with kl_scale(n_samples):
```
Comment (ronshiff1) [inline pybandits/model.py]: I don't understand why this wraps the sampling code...

File: pybandits/model.py [hunk 3]
```diff
@@ -1454,7 +1521,42 @@ def __setstate__(self, state: dict) -> None:
...
+            Number of data points in the current batch. Used to
+            compute the data-dependent part of the scale factor.
+
+        Returns
+        -------
+        contextlib.AbstractContextManager
+            Either numpyro.handlers.scale(scale=beta) or nullcontext().
+        """
+        kl_tau = self._update_kwargs.get("kl_tau")
+        if kl_tau is not None:
```
Comment (ronshiff1) [inline pybandits/model.py]: where do we take into account:
self.model_params.embedding_params.embeddings?

File: pybandits/model.py
Comment (ronshiff1) [inline pybandits/model.py:1669]: I think it's a good idea to validate that update_kwargs values are defined correctly 
(for example, that num_particles is a natural number)
maybe we can do something like:
```
from pydantic import PositiveInt, ValidationError
class ViUpdateKwargs(PyBanditsBaseModel):
    num_steps: PositiveInt = 1000
    method: VIMethods = "advi"
    optimizer_type: str = "sgd"
    optimizer_kwargs: dict = Field(default_factory=lambda: {"step_size": 0.01})
    batch_size: Optional[PositiveInt] = None
    num_particles: PositiveInt = 1
    # ... other VI keys ...
# inside _arrange_update_kwargs, after merge:
try:
    validated = ViUpdateKwargs.model_validate(self.update_kwargs)
except ValidationError as e:
    raise ValueError(f"Invalid VI update_kwargs: {e}") from e
self.update_kwargs = validated.model_dump()
```

**Review by ronshiff1 — APPROVED** [review]
